### DISCUSSION TITLE
Updating Async Atoms that make XHR requests

### DISCUSSION THREAD

**Discussion by sazzer** [pmndrs/jotai#1458] [Q&A] [answered]
Hi there,

I've got an Async Atom that makes XHR requests by means of Ketting. This works great for read-only atoms, but I can't work out how to update them.

The read-only atom looks like this:
```
export const currentUserAtom = atom(async (get) => {
  const home = await get(apiHomeAtom);

  const userResource = await home.follow<UserProperties>('authenticated-as');
  const user = await userResource.get();

  return user;
});
```

And then anything that uses this has access to the user resource. If it's not already loaded then it will be, and all is good.

What I want now is to be able to update the user. Adding a setter method works as expected - I can call it using the standard means and it makes the correct XHR calls:
```
export const currentUserAtom = atom(
  async (get) => {
    const home = await get(apiHomeAtom);

    const userResource = await home.follow<UserProperties>('authenticated-as');
    const user = await userResource.get();

    return user;
  },
  async (get, set, values) => {
    const user = await get(currentUserAtom);

    const editAction = user.action('default');
    const updated = await editAction.submit(values);

    // What now?
  }
);
```

What I can't work out is how to then get the atom state to update with the result of this XHR call.

What I've tried so far:
* Returning the updated user - has no effect
* Calling `set(currentUserAtom, updated)` - gives an error "atom not writable" because I'm updating the same one I'm calling from.
* Adding another atom that I can use to trigger a reset of this one. This works, but it causes the app to make another GET call to reload the user, instead of just using the response of the PUT call.

Is there some correct way to achieve this? Or is this the best we've got so far?

Cheers

**Comment by dai-shi** [comment, marked as answer]
Maybe `atomWithDefault` util helps.

**Reply by sazzer** [reply]
That actually works remarkably well.

I've just implemented it, pulling the update calls out of the atom itself and into my custom hook that wraps it. I'm now seeing a single network call to `PUT /users/id`, after which the atom state updates and the UI updates to reflect this. :)

Whether this is the *correct* answer or not, it certainly does what I need :)

**Reply by sazzer** [reply]
For reference, this is what it all looks like now:
```
import { useAtom } from 'jotai';
import { atomWithDefault } from 'jotai/utils';
import { apiHomeAtom } from '../home';
import { newUser, User, UserProperties } from './user-entity';

/** Atom to represent the current user */
export const currentUserAtom = atomWithDefault(async (get) => {
  const home = await get(apiHomeAtom);

  const userResource = await home.follow<UserProperties>('authenticated-as');
  const user = await userResource.get();

  return user;
});

/** Extension to the User that allows changes to be saved */
export type SaveableUser = User & {
  update(values: UserProperties): Promise<void>;
};

/** The shame of the Current User hook */
export interface CurrentUserHook {
  user: SaveableUser;
}

/**
 * Hook to get the current user details.
 */
export function useCurrentUser(): CurrentUserHook {
  const [user, setUser] = useAtom(currentUserAtom);

  const userEntity = newUser(user);

  return {
    user: {
      async update(values) {
        const editAction = user.action('default');
        const updated = await editAction.submit(values);

        setUser(updated);
      },
      ...userEntity
    }
  };
}
```

Where the `User` and `newUser` portions are just a domain entity wrapper around the Ketting state to allow other areas to work in domain terms.
